Kta 60 results after uploading data

sandoche opened the pull request:
A PR related to KTA-60 (check it out for more infos) but in short:
- Adding the results at the end of the flow
- Possibility to restart if it failed
- Changed a bit the UI to make the full flow more consistent
- Improved a bit the UX
- Refactored the code (adding types, services, more components, file structure)

There is still something not very clear to know when to show the loading animation, and when to show the failing / restart message that is why I sent a message to the Onfido support.

Depending on the results I may have to update:
- https://github.com/NEARFoundation/near-kyc/blob/KTA-60-results-after-uploading-data/pages/%5Bkey%5D/results.tsx
- https://github.com/NEARFoundation/near-kyc/blob/KTA-60-results-after-uploading-data/pages/api/check-results.ts#L24 (and the response  instead of `status: simplifiedStatus.get(check.status) ?? CheckResultsStatus.finished,` maybe I would need to send `status: simplifiedStatus.get(check.status) ?? CheckResultsStatus.willTakeLonger`)

I will later write some E2E testings.

ryancwalsh commented on pages/[key]/index.tsx:
```diff
@@ -104,6 +80,33 @@ const StartPage: NextPage = () => {
+
+      const applicantProperties = JSON.parse(localStorageUserData) as ApplicantProperties;
+
+      if (!applicantProperties.dob || !applicantProperties.email || !applicantProperties.firstName || !applicantProperties.lastName) {
```
If we're able to (if it's not a field name defined by OnFido), I'd prefer not to use abbreviations like "dob". See https://github.com/NEARFoundation/near-kyc/blob/c44fa7cb75f307a60900f17aef12b65e9351257d/types/ApplicantProperties.ts and https://github.com/NEARFoundation/near-kyc/pull/10/files#diff-06066658bada731114226fdb8eee079b437ac8d1888348aa566d324a7212897cR35

Let's call it something clearer, like birthDate.

sandoche replied:
Sadly the dob is from Onfido: https://documentation.onfido.com/#applicants

ryancwalsh commented on components/results/ResultsError.tsx:
```diff
@@ -0,0 +1,11 @@
+
+export default function ResultsFailure(): JSX.Element {
+  return (
+    <CenteredCardContent title="An error occured" description="Sorry an error occured, we invite you to try again." iconClasses="fa fa-exclamation-circle text-warning mb-4">
```
Change to:

> Sorry, an error occurred; we invite you to try again.

sandoche replied:
Changed, thanks!

ryancwalsh commented on components/results/ResultsFailure.tsx:
```diff
@@ -0,0 +1,11 @@
+
+export default function ResultsFailure(): JSX.Element {
+  return (
+    <CenteredCardContent title="Verification failed" description="Sorry, we invite you to try again the verification." iconClasses="fa fa-times-circle text-danger mb-4">
```
What's the difference between ResultFailure and ResultsError? https://github.com/NEARFoundation/near-kyc/pull/10/files#r990093777

The error messages look almost the same but not quite. Should they be identical?

Oh, I see that the titles differ.

So maybe the message in this one should say:

> We could not verify your identity. If you believe that you submitted any information incorrectly, you may try again.

or something like that.

sandoche replied:
Error is more due to a bug user do not exist, or backend error.

Failure is when the validation didn't pass

sandoche replied:
Thanks for your text it's indeed more clear!

ryancwalsh commented on components/results/ResultsLoading.tsx:
```diff
@@ -0,0 +1,22 @@
+      description={
+        willTakeLonger ? (
+          <p>
+            <strong>This verification takes more time than usual.</strong>
```
Change to:

> This identity verification is taking more time than usual.

sandoche replied:
Good catch I am ashamed that I wrote that so wrong 😅
I guess I didn't re-read what I wrote. Thanks

ryancwalsh replied:
I'm always impressed by you (and others) who speak multiple languages. It's amazing. I certainly wouldn't be embarrassed.

ryancwalsh commented on components/results/ResultsSuccess.tsx:
```diff
@@ -0,0 +1,11 @@
+  return (
+    <CenteredCardContent
+      title="Verification validated"
+      description="You can now close this window. The NEAR Foundation will contact you for the next steps."
```
This feels anti-climatic. We just asked someone to go through a bunch of hassle, and the result came back positive, so I think we should say something happier and more positive, especially since we're then bringing them the bad news that they have to wait longer on something else.

Maybe we want to say:

> Congratulations! 🎉 We have verified your identity and are excited to move forward. NEAR Foundation will contact you for the next steps shortly. You can now close this window.

sandoche replied:
Good one! Thanks a lot!

ryancwalsh commented on constants/index.tsx:
```diff
@@ -0,0 +1,5 @@
+export const COOKIE_CHECK_ID_NAME = 'onfido-check-id';
+export const LOCALSTORAGE_USER_DATA_NAME = 'onfido-user-data';
+export const SHORT_POLLING_INTERVAL = 1000;
+export const LONG_POLLING_INTERVAL = 30000;
```
Presumably 30_000 here means 30 seconds, but I'd clarify these lines with comments. And then maybe describe the difference between when SHORT_ or LONG_ would be used.

sandoche replied:
Thanks done!

ryancwalsh commented on pages/api/check-results.ts:
```diff
@@ -0,0 +1,40 @@
+  const check = await onfido.check.find(checkId);
+
+  if (!checkId || !check) {
+    res.status(404).json({ isClear: null, status: CheckResultsStatus.notFound });
```
To avoid https://en.wikipedia.org/wiki/Magic_number_(programming), I'm a fan of pulling these out like https://github.com/NEARFoundation/tx-tracking-app/blob/7ba69e69b6d90fbf999854b9c04e377c2dc21979/shared/helpers/statusCodes.ts

sandoche replied:
Fixed and linting rule added!

ryancwalsh commented on pages/api/check-results.ts:
```diff
@@ -0,0 +1,40 @@
+  ]);
+
+  // This need to be updated / improved depending on the answer from the support
+  res.status(200).json({
```
See https://github.com/NEARFoundation/near-kyc/pull/10/files#r990107300

ryancwalsh commented on pages/api/create-check.ts:
```diff
@@ -39,6 +36,9 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse<
+    res.setHeader('Set-Cookie', `${COOKIE_CHECK_ID_NAME}=${check.id}; Max-Age=${COOKIES_EXPIRATION_TIME}; Path=/`);
+
     console.log('Returning result', endpointName);
     res.status(200).json(check);
```
See https://github.com/NEARFoundation/near-kyc/pull/10/files#r990107300

ryancwalsh reviewed (approved):
I added some comments but nothing that stops me from approving.

Also, whenever you're submitting something that you want me to consider more carefully, either include screenshots and/or a 30-second video showing me what you want me to see, or tell me to build everything and run it myself (which I didn't do here since I assumed you'd tell me if anything needs that level of attention).

sandoche commented:
- [ ] Correct this: https://github.com/NEARFoundation/near-kyc/pull/11#discussion_r990084735

sandoche commented:
@ryancwalsh just for your information: I updated the check state I was looking at the wrong place in the doc, thanks to the support I found this out, so I fixed it in this commit: https://github.com/NEARFoundation/near-kyc/pull/10/commits/ee5c5097a8c5b6fd483c786c2365d25b45332066#diff-1130d586ce172f3f2f257e9b1875bd308f3335817f0a91f4df1d6ac8a93e101d

ryancwalsh commented on .eslintrc.js:
```diff
@@ -34,7 +34,8 @@ module.exports = {
     'simple-import-sort/exports': 'error',
-    '@typescript-eslint/prefer-nullish-coalescing': 'warn',
+    '@typescript-eslint/prefer-nullish-coalescing': 'error',
+    'no-magic-numbers': 'error',
```
Oh cool, I didn't know about this rule. 👍  I added to https://near-foundation.atlassian.net/jira/software/c/projects/ENG/boards/1?modal=detail&selectedIssue=ENG-156&quickFilter=1
